lib/tbot/service/service.go

```diff
@@ -0,0 +1,163 @@
+}
+
+// NewService creates a service with the given name and handler.
+func NewService[HandlerT Handler](name string, handler HandlerT) *Service[HandlerT] {
+	return &Service[HandlerT]{
+		name:    name,
+		handler: handler,
+		status:  NewWatchedValue(status.Initializing),
+	}
+}
+
+// Service implements a user-configurable part of tbot, such as generating AWS
+// credentials, serving the SPIFFE Workload API, or proxying database traffic.
+//
+// Services have a status that describes whether they're currently healthy.
+// Dependent services can subscribe to status changes using WatchStatusChanges
+// or wait for the service to have a given status using WaitForStatus.
+//
+// If the service's handler implements OneShotHandler, it can be used with the
+// `--one-shot` flag.
+//
+// The service lifecycle is managed by a Supervisor.
+type Service[HandlerT Handler] struct {
+	name    string
+	handler HandlerT
+	status  *WatchedValue[status.Status]
+}
+
+// Status returns the service's current status.
+func (s *Service[HandlerT]) Status() status.Status {
+	return s.status.Get()
+}
+
+// ErrWrongStatus is returned from Service.Wait and Service.WaitForStatus when
+// the service is not in the required status.
+var ErrWrongStatus = errors.New("supervisor: service has wrong status")
+
+// Wait blocks until the service's status is Ready and then returns the handler.
+//
+// If the given context is canceled or reaches its deadline, Wait will unblock
+// immediately and return the context's error. This is useful for implementing
+// timeout/fallback behavior.
+//
+// When called from another service running in one-shot mode, Wait will block
+// until the service's status changes *for the first time* only (because in
+// one-shot mode an unhealthy service wouldn't ever recover). If the service
+// isn't ready, ErrWrongStatus will be returned.
+func (s *Service[HandlerT]) Wait(ctx context.Context) (HandlerT, error) {
+	return s.WaitForStatus(ctx, status.Ready)
+}
+
+// WaitForStatus blocks until the service's status matches one of the given
+// statuses and then returns the handler.
+//
+// If the given context is canceled or reaches its deadline, WaitForStatus will
+// unblock immediately and return the context's error. This is useful for
+// implementing timeout/fallback behavior.
+//
+// When called from another service running in one-shot mode, WaitForStatus will
+// block until the service's status changes *for the first time* only (because
+// in one-shot mode an unhealthy service wouldn't ever recover). If the service's
+// status doesn't match one of the expected statuses, ErrWrongStatus will be returned.
+func (s *Service[HandlerT]) WaitForStatus(ctx context.Context, statuses ...status.Status) (HandlerT, error) {
+	current, watcher := s.status.Watch()
+	defer watcher.Close()
+
+	if slices.Contains(statuses, current) {
+		return s.handler, nil
+	}
+
+	var zero HandlerT
+	if isOneShot(ctx) && current != status.Initializing {
```

Comment:
Is it possible to assert s.handler.(OneShotHandler) instead of relying on the context key?

Reply:
Handlers that implement `OneShotHandler` can also be run as long-running services by calling `Run`.

I realised the behavior here wasn't quite right, though. What I was trying to achieve was unblocking immediately once the `OneShot` method had run (because the service's status would never change after that) but this implementation assumed that the caller was also within a `OneShot` method.

Instead, in 21f620c6a8318ab8cd202128f219b9cab43b3b28 I've made it so the supervisor closes a channel once the handler (long-running or one-shot) has totally finished, so we know definitely that the service's status will never change again.